_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,871 @@
+
+Regardless, the very top of the stack frame will _always_ contain the same set of registers. We can represent this with C code as:
+
+```c
+typedef struct __attribute__((packed)) ContextStateFrame {
+  uint32_t r0;
+  uint32_t r1;
+  uint32_t r2;
+  uint32_t r3;
+  uint32_t r12;
+  uint32_t lr;
+  uint32_t return_address;
+  uint32_t xpsr;
+} sContextStateFrame;
+```
+
+This register state will be pushed to the stack which was active prior to exception entry. ARM Cortex-M devices have [two stack pointers](https://interrupt.memfault.com/blog/cortex-m-rtos-context-switching#stack-pointers-and-usage), `msp` & `psp`. Upon exception entry, the active stack pointer (`CONTROL.SPSEL`) is encoded in bit 2 of the `EXC_RETURN` value pushed to the link register. If the bit is set, the `psp` was active, else the `msp` was active. We can therefore put together a small shim assembly function that figures out the current sp (whose current layout will match `sContextStateFrame`) and pass that to a C function for further processing.
+
+```c
+#define HARDFAULT_HANDLING_ASM(_x)               \
+  __asm volatile(                                \
+      "tst lr, #4 \n"                            \
+      "ite eq \n"                                \
+      "mrseq r0, msp \n"                         \
+      "mrsne r0, psp \n"                         \
+      "b my_fault_handler_c \n"                  \
+                                                 )
+```
+
+The `my_fault_handler_c` function will usually look something like this:
+
+```c
+// NOTE: If you are using CMSIS, the registers can also be accessed through
+// CoreDebug->DHCSR & CoreDebug_DHCSR_C_DEBUGEN_Msk
+#define HALT_IF_DEBUGGING()                              \
+  do {                                                   \
+    if ((*(volatile uint32_t *)0xE000EDF0) & (1 << 0)) { \
+      __asm("bkpt 1");                                   \
+    }                                                    \
+  } while (0)
+
+// Disable optimizations for this function so "frame" argument does not get optimized away
+__attribute__((optimize("O0")))
+void my_fault_handler_c(sContextStateFrame *frame) {
+  // If and only if a debugger is attached, execute a breakpoint instruction so
+  // we can take a look at what triggered the fault
+  HALT_IF_DEBUGGING();
+
+  // Logic for dealing with the exception. Typically:
+  //  - log the fault which occurred for postmortem analysis
+  //  - If the fault is recoverable,
+  //    - clear errors and return back to Thread Mode
+  //  - else
+  //    - reboot system
+}
+```
+
+## Recovering From A Fault
```

Comment:
I'd move the "Recovering" until after the "Automating" section so the narrative is more compelling.

Doing by hand -> adding some automation -> what if we don't want to reset?